File: internal/probe/http/http.go
package httpprobe

import (
	"context"
	"fmt"
	"net"
	"net/http"
	"time"

	"icurl/internal/evidence"
	"icurl/internal/probe"
	"icurl/internal/request"
)

type Runner interface {
	Do(context.Context, request.Config) (request.Result, error)
}

type Probe struct {
	Runner Runner
}

func (p Probe) Run(ctx context.Context, target probe.Target) evidence.ProbeResult {
	started := time.Now()
	result := evidence.ProbeResult{
		ProbeName:  "HTTP",
		Layer:      evidence.LayerHTTP,
		Target:     target.URL.String(),
		StartedAt:  &started,
		Confidence: evidence.ConfidenceObserved,
	}

	if p.Runner == nil {
		finish(&result, evidence.ResultFailed, evidence.ErrorUnknown, "HTTP runner is nil")
		return result
	}

	urlStr := target.URL.String()
	if len(target.IPs) > 0 {
		// Replace hostname with IP to skip DNS resolution
		u := *target.URL
		u.Host = net.JoinHostPort(target.IPs[0].String(), target.Port)
		urlStr = u.String()
	}

	response, err := p.Runner.Do(ctx, request.Config{
		URL:            urlStr,
		Host:           target.Host, // Pass explicit host for SNI
		Method:         http.MethodHead,
		Protocol:       request.ProtocolAuto,
		MaxTime:        10 * time.Second,
		ConnectTimeout: 5 * time.Second,
	})
	if err != nil {
		finish(&result, evidence.ResultFailed, evidence.ErrorProtocol, err.Error())
		return result
	}

	if response.StatusCode >= 400 {
		result.Observations = []string{"HTTP responded with " + response.Protocol}
		finish(&result, evidence.ResultFailed, evidence.ErrorProtocol, fmt.Sprintf("HTTP %d", response.StatusCode))
		return result
	}

	result.Observations = []string{fmt.Sprintf("HTTP %d over %s", response.StatusCode, response.Protocol)}
	finish(&result, evidence.ResultOK, evidence.ErrorNone, "")
	return result
}

func finish(result *evidence.ProbeResult, probeResult evidence.Result, kind evidence.ErrorKind, message string) {
	finished := time.Now()
	result.FinishedAt = &finished
	result.DurationNS = finished.Sub(*result.StartedAt).Nanoseconds()
	result.Result = probeResult
	result.ErrorKind = kind
	result.ErrorMessage = message
}

File: icurl-rs/src/probe/http.rs
use crate::evidence::{classify_error_string, ErrorKind, Layer, ProbeResult, ProbeResultStatus};
use crate::probe::{BoxFuture, Probe, Target};
use crate::request::{Config as ReqConfig, Protocol as ReqProtocol, execute as request_execute};
use std::time::Duration;

pub struct HttpProbe;

impl Probe for HttpProbe {
    fn probe_name(&self) -> &'static str { "HTTP" }

    fn run<'a>(&'a self, target: &'a Target) -> BoxFuture<'a, ProbeResult> {
        Box::pin(async move {
            let mut result = ProbeResult::new("HTTP", Layer::Http, target.url.as_ref());

            let mut response = None;
            let mut last_err = None;

            let ips = &target.ips;
            if ips.is_empty() {
                let req_cfg = ReqConfig {
                    url: target.url.to_string(),
                    host: target.host.clone(),
                    method: "HEAD".to_string(),
                    protocol: ReqProtocol::Auto,
                    max_time: Duration::from_secs(10),
                    connect_timeout: Duration::from_secs(5),
                    ..Default::default()
                };
                match request_execute(&req_cfg).await {
                    Ok(res) => response = Some(res),
                    Err(e) => last_err = Some(e.to_string()),
                }
            } else {
                for ip in ips {
                    let mut u = target.url.clone();
                    let _ = u.set_host(Some(&ip.to_string()));
                    let _ = u.set_port(Some(target.port));

                    let req_cfg = ReqConfig {
                        url: u.to_string(),
                        host: target.host.clone(),
                        method: "HEAD".to_string(),
                        protocol: ReqProtocol::Auto,
                        max_time: Duration::from_secs(10),
                        connect_timeout: Duration::from_secs(5),
                        ..Default::default()
                    };
                    match request_execute(&req_cfg).await {
                        Ok(res) => {
                            response = Some(res);
                            break;
                        }
                        Err(e) => {
                            last_err = Some(e.to_string());
                        }
                    }
                }
            }

            if let Some(res) = response {
                if res.status_code >= 400 {
                    result.add_observation(&format!("HTTP responded with {}", res.protocol));
                    result.finish(
                        ProbeResultStatus::Failed,
                        ErrorKind::Protocol,
                        &format!("HTTP {}", res.status_code),
                    );
                } else {
                    result.add_observation(&format!("HTTP {} over {}", res.status_code, res.protocol));
                    result.finish(ProbeResultStatus::Ok, ErrorKind::None, "");
                }
            } else {
                let err_str = last_err.unwrap_or_else(|| "HTTP request failed".to_string());
                let (status, kind) = classify_error_string(&err_str);
                result.finish(status, kind, &err_str);
            }

            result
        })
    }
}
